### PR TITLE
[FLINK-20522][table] Make implementing a built-in function straightforward

### PR DESCRIPTION
**Pull request by twalthr**
## What is the purpose of the change

This PR completes the envisioned design of FLIP-32 and FLIP-65. It makes implementing built-in
functions as easy as implementing user-defined functions while keeping function definition and body (aka runtime
implementation) separately.

It introduces the concept of `SpecializedFunctions` to hook into the planning phase for generating
runtime implementation lazily.

For proving the concepts, the PR implements two frequently requested functions: `TYPEOF()` and `IFNULL()`

With this PR, adding a built-in function requires changes in only 2 classes: `BuiltInFunctionDefinitions` for the definition and the class that contains the actual runtime logic.

## Brief change log

See commit messages.

## Verifying this change

This change added tests and can be verified as follows:
- `FunctionITCase`
- `MiscFunctionsITCase`

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn/Mesos, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? yes
  - If yes, how is the feature documented? docs

### CODE REVIEW

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+ */
+@PublicEvolving
+public interface SpecializedFunction extends FunctionDefinition {
+
+	/**
+	 * Provides a runtime implementation that is specialized for the given call and session.
+	 *
+	 * <p>The method must return an instance of {@link UserDefinedFunction} or throw a {@link TableException}
+	 * if the given call is not supported. The returned instance must have the same {@link FunctionDefinition}
+	 * semantics but can have a different {@link #getTypeInference(DataTypeFactory)} implementation.
```
Comment (dawidwys) [inline flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java:49]: I am wondering if we should be more elaborate here and add a more prominent warning.

Can the type inference change arbitrarily? I think it is quite easy to abuse it, e.g. :
```
	/**
	 * A specialized "compile time" function for returning the argument's data type.
	 */
	public static class NumberScalarFunction extends ScalarFunction implements SpecializedFunction {

		private final boolean doubleNumber;

		public NumberScalarFunction() {
			this.doubleNumber = false;
		}

		public NumberScalarFunction(boolean doubleNumber) {
			this.doubleNumber = doubleNumber;
		}

		@SuppressWarnings("unused")
		public Object eval(@DataTypeHint(inputGroup = InputGroup.ANY) Object unused) {
			if (doubleNumber) {
				return 0D;
			} else {
				return 0;
			}
		}

		@Override
		public TypeInference getTypeInference(DataTypeFactory typeFactory) {
			return TypeInference.newBuilder()
				.inputTypeStrategy(InputTypeStrategies.WILDCARD)
				.outputTypeStrategy(doubleNumber ? TypeStrategies.explicit(DataTypes.DOUBLE()) : TypeStrategies.explicit(DataTypes.INT()))
				.build();
		}

		@Override
		public NumberScalarFunction specialize(SpecializedContext context) {
			if (context.getCallContext().isArgumentLiteral(0) && context
				.getCallContext()
				.getArgumentValue(0, Boolean.class)
				.get()) { // this could be based e.g. on the input type
				return new NumberScalarFunction(true);
			} else {
				return new NumberScalarFunction(false);
			}
		}
	}

	public static class DoubleScalarFunction extends ScalarFunction {

		public String eval(double unused) {
			return Double.toString(unused);
		}

		public int eval(int number) {
			return number;
		}

		@Override
		public TypeInference getTypeInference(DataTypeFactory typeFactory) {
			Map<InputTypeStrategy, TypeStrategy> map = new HashMap<>();
			map.put(
				InputTypeStrategies.sequence(InputTypeStrategies.logical(LogicalTypeRoot.DOUBLE)),
				TypeStrategies.explicit(DataTypes.DOUBLE())
			);
			map.put(
				InputTypeStrategies.sequence(InputTypeStrategies.logical(LogicalTypeRoot.INTEGER)),
				TypeStrategies.explicit(DataTypes.INT())
			);
			return TypeInference.newBuilder()
				.inputTypeStrategy(InputTypeStrategies.sequence(InputTypeStrategies.or(
					InputTypeStrategies.logical(LogicalTypeRoot.DOUBLE),
					InputTypeStrategies.logical(LogicalTypeRoot.INTEGER)
				)))
				.outputTypeStrategy(
					TypeStrategies.mapping(
						map
					)
				)
				.build();
		}
	}

And then: 

		tEnv().createTemporarySystemFunction("Innner", NumberScalarFunction.class);
		tEnv().createTemporarySystemFunction("Outerr", DoubleScalarFunction.class);

		final TableResult result = tEnv()
			.executeSql(
				"SELECT "
					+ " Outerr(Innner(s)) "
					+ "FROM SourceTable");
```

The end result type will be `INT` though some invokations of the inner function might produce double type.

Correct me if I am wrong but the specialized function should have a `TypeInference` that is a specialized version of the original one.

In short I just thought it would be better to make it very clear this is a rather low level and very powerful kind of function that requires a lot of thought.

**Review by dawidwys — APPROVED** [review]
Overall +1 looks really good!

I had two rather minor suggestions.

File: flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java (hunk 1 above)
Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java:49]: I cannot reproduce your concern. Given the following simplified code:
```
    @Test
    public void testSpecializedFunction() {
        final List<Row> sourceData =
                Arrays.asList(
                        Row.of("Bob", 1, new BigDecimal("123.45")),
                        Row.of("Alice", 2, new BigDecimal("123.456")));

        TestCollectionTableFactory.reset();
        TestCollectionTableFactory.initData(sourceData);

        tEnv().executeSql(
                        "CREATE TABLE SourceTable("
                                + "  s STRING, "
                                + "  i INT,"
                                + "  d DECIMAL(6, 3)"
                                + ")"
                                + "WITH ("
                                + "  'connector' = 'COLLECTION'"
                                + ")");

        tEnv().createTemporarySystemFunction("Innner", NumberScalarFunction.class);
        tEnv().createTemporarySystemFunction("Outerr", DoubleScalarFunction.class);

        final TableResult result =
                tEnv().executeSql("SELECT " + " Outerr(Innner(s)) " + "FROM SourceTable");
    }

    /** A specialized "compile time" function for returning the argument's data type. */
    public static class NumberScalarFunction extends ScalarFunction implements SpecializedFunction {

        private final boolean doubleNumber;

        public NumberScalarFunction() {
            this.doubleNumber = false;
        }

        public NumberScalarFunction(boolean doubleNumber) {
            this.doubleNumber = doubleNumber;
        }

        @SuppressWarnings("unused")
        public Object eval(@DataTypeHint(inputGroup = InputGroup.ANY) Object unused) {
            if (doubleNumber) {
                return 0D;
            } else {
                return 0;
            }
        }

        @Override
        public TypeInference getTypeInference(DataTypeFactory typeFactory) {
            return TypeInference.newBuilder()
                    .outputTypeStrategy(
                            doubleNumber
                                    ? TypeStrategies.explicit(DataTypes.DOUBLE())
                                    : TypeStrategies.explicit(DataTypes.INT()))
                    .build();
        }

        @Override
        public NumberScalarFunction specialize(SpecializedContext context) {
            return new NumberScalarFunction(true);
        }
    }

    public static class DoubleScalarFunction extends ScalarFunction {

        public String eval(double unused) {
            return Double.toString(unused);
        }

        public int eval(int number) {
            return number;
        }

        @Override
        public TypeInference getTypeInference(DataTypeFactory typeFactory) {
            Map<InputTypeStrategy, TypeStrategy> map = new HashMap<>();
            map.put(
                    InputTypeStrategies.sequence(
                            InputTypeStrategies.logical(LogicalTypeRoot.DOUBLE)),
                    TypeStrategies.explicit(DataTypes.DOUBLE()));
            map.put(
                    InputTypeStrategies.sequence(
                            InputTypeStrategies.logical(LogicalTypeRoot.INTEGER)),
                    TypeStrategies.explicit(DataTypes.INT()));
            return TypeInference.newBuilder()
                    .inputTypeStrategy(
                            InputTypeStrategies.sequence(
                                    InputTypeStrategies.or(
                                            InputTypeStrategies.logical(LogicalTypeRoot.DOUBLE),
                                            InputTypeStrategies.logical(LogicalTypeRoot.INTEGER))))
                    .outputTypeStrategy(TypeStrategies.mapping(map))
                    .build();
        }
    }
```

Comment (twalthr) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java:49]: The following exception message would be thrown:
```
org.apache.flink.table.planner.codegen.CodeGenException: Mismatch of expected output data type 'INT' and function's output type 'DOUBLE'.

	at org.apache.flink.table.planner.codegen.calls.BridgingFunctionGenUtil$.verifyOutputType(BridgingFunctionGenUtil.scala:369)
	at org.apache.flink.table.planner.codegen.calls.BridgingFunctionGenUtil$.verifyFunctionAwareOutputType(BridgingFunctionGenUtil.scala:357)
	at org.apache.flink.table.planner.codegen.calls.BridgingFunctionGenUtil$.generateFunctionAwareCallWithDataType(BridgingFunctionGenUtil.scala:101)
	at org.apache.flink.table.planner.codegen.calls.BridgingFunctionGenUtil$.generateFunctionAwareCall(BridgingFunctionGenUtil.scala:65)
	at org.apache.flink.table.planner.codegen.calls.BridgingSqlFunctionCallGen.generate(BridgingSqlFunctionCallGen.scala:73)
```

Comment (dawidwys) [reply flink-table/flink-table-common/src/main/java/org/apache/flink/table/functions/SpecializedFunction.java:49]: Sorry, my example was wrong, in a respect that I did not use a literal and therefore never ended up with `new NumberScalarFunction(true)`. Because of the example I assumed we do not have such check, but obviously we do. Again sorry for the confusion. Nice that we have that check!

Still it won't harm to have a slightly more prominent warning.

**Comment by twalthr** [comment]
Thanks for the detailed review @dawidwys. I addressed your comments and rebased the PR. I will merge this once Azure gives green light.
